File: BFS.cpp
#include<bits/stdc++.h>
typedef long long ll;
typedef unsigned long long ull;
using namespace std;


void bfs(vector<ll> adjList[], vector<ll> &distance, vector<ll> &parent, ll source)
{
    queue<ll> q;
    q.push(source);
    distance[source] = 0;

    while (!q.empty())
    {
        ll u = q.front();
        q.pop();

        for(int i=0; i < adjList[u].size(); i++)
        {
            int v = adjList[u][i];

            if(distance[v] == -1)
            {
                distance[v] = distance[u] + 1;
                parent[v] = u;
                q.push(v);
            }
        }
    }
}


int main()
{
    ios_base::sync_with_stdio(false); cin.tie(nullptr);
//    memset(dp, -1, sizeof(dp));
    ll numOfNodes, numOfEdges; cin>>numOfNodes>>numOfEdges;
    vector<ll> adjList[numOfNodes+1];
    vector<ll> distance(numOfNodes+1, -1);
    vector<ll> parent(numOfNodes+1, -1);

    for(int i=1; i<=numOfEdges; i++)
    {
        ll u, v; cin>>u>>v;
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }

    ll source; cin>>source;

    bfs(adjList, distance, parent, source);

    int destination, copyDestination; cin>>destination; copyDestination = destination;
    vector<int> result;
    while(parent[destination] != source)
    {
        result.push_back(parent[destination]);
        destination = parent[destination];
    }
    reverse(result.begin(), result.end());

    cout<<source<<" ";
    for(int i=0; i<result.size(); i++)
    {
        cout<<result[i]<<" ";
    }
    cout<<copyDestination<<" ";

}

File: BFS.py
def bfs(adjList, distance, parent, source):
    q = []  # a queue
    q.append(source)
    distance[source] = 0  # source distance is ZERO

    while q:  # queue not equals to empty
        u = q.pop(0)

        for i in range(len(adjList[u])):
            v = adjList[u][i]

            if distance[v] == -1:
                distance[v] = distance[u] + 1
                parent[v] = u
                q.append(v)


# driver code / Global code

numOfNodes, numOfEdges = map(int, input().split())

adjList = [[] for _ in range(numOfNodes + 1)]
distance = [-1 for _ in range(numOfNodes + 1)]
parent = [-1 for _ in range(numOfNodes + 1)]


for i in range(numOfEdges):
    u, v = map(int, input().split())
    adjList[u].append(v)
    adjList[v].append(u)

source = int(input())

bfs(adjList, distance, parent, source)


# printing source to destination path

destination = int(input())
copyDestination = destination
result = []

while parent[destination] != source:
    result.append(parent[destination])
    destination = parent[destination]

result = reversed(result)
print(source, end=" ")
for i in result:
    print(f'{i} ', end="")
print(copyDestination)
